File: java/src/main/structural/adapter/Adapter.java
package structural.adapter;

public class Adapter implements Target {
    private final Adaptee adaptee;

    public Adapter(Adaptee adaptee) {
        this.adaptee = adaptee;
    }

    @Override
    public void request() {
        // Delegates the call to the Adaptee's specificRequest method
        System.out.println("Adapter: Translating request...");
        adaptee.specificRequest();
    }
}

// Example Usage
// public class Main {
//     public static void main(String[] args) {
//         Adaptee adaptee = new Adaptee();
//         Target adapter = new Adapter(adaptee);
//
//         // Client code works with the Target interface
//         adapter.request();
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Ignoring Interface Requirements:
//    - Mistake: Not implementing the entire Target interface.
//    - Fix: Ensure the Adapter class fully implements all methods of the Target interface.
//
// 2. Tight Coupling:
//    - Mistake: Directly modifying the Adaptee instead of adapting it.
//    - Fix: Always use the Adapter to mediate between the Target and Adaptee.
//
// 3. Overcomplicating the Adapter:
//    - Mistake: Adding unnecessary logic to the Adapter.
//    - Fix: Keep the Adapter focused on translation logic only.
//
// 4. Assuming Compatibility Without Testing:
//    - Mistake: Assuming the Adaptee will work seamlessly after adaptation.
//    - Fix: Test the Adapter thoroughly with the Target interface.

File: python/src/structural/adapter/adapter.py
from src.structural.adapter.target import Target
from src.structural.adapter.adaptee import Adaptee


class Adapter(Target):
    def __init__(self, adaptee: Adaptee):
        self._adaptee = adaptee

    def request(self) -> str:
        return (
            f"Adapter: (TRANSLATED) {self._adaptee.specific_request()[::-1]}"
            )

# Example Usage:
# if __name__ == "__main__":
#     adaptee = Adaptee()
#     target = Adapter(adaptee)
#     print(target.request())
#
# Output:
# Adapter: (TRANSLATED) Special behavior of the Adaptee.

# Common Mistakes to Avoid:
#
# 1. Misunderstanding Adapter's Role:
#    - Mistake: Implementing business logic inside the adapter.
#    - Fix: Keep the adapter focused only on translating interfaces.
#
# 2. Tight Coupling to the Adaptee:
#    - Mistake: Directly exposing Adaptee methods in the Adapter.
#    - Fix: Hide adaptee-specific logic from the client interface.
#
# 3. Violating Liskov Substitution Principle:
#    - Mistake: Making Adapter behave differently than Target.
#    - Fix: Ensure Adapter is substitutable for Target without side effects.
#
# 4. Ignoring Composition:
#    - Mistake: Using inheritance when composition is more appropriate.
#    - Fix: Favor composition over inheritance for more flexibility.
